lib/bosh/BoshDirectorClient.js

```diff
@@ -16,48 +16,139 @@ const NotFound = errors.NotFound;
 const BadRequest = errors.BadRequest;
 
 class BoshDirectorClient extends HttpClient {
-  constructor(boshDirectorConfig) {
-    const boshDirector = boshDirectorConfig || config.director;
+  constructor() {
     super({
-      baseUrl: boshDirector.url,
-      auth: {
-        user: boshDirector.username,
-        pass: boshDirector.password
-      },
       headers: {
         Accept: 'application/json,text/plain;q=0.9'
       },
-      followRedirect: false,
-      rejectUnauthorized: !boshDirector.skip_ssl_validation
+      followRedirect: false
+    });
+    // activeConfigs - directors supporting lifecycle oprations
+    this.activeConfigs = BoshDirectorClient.getPrimaryConfigs();
+    // activePrimary - directors supporting 'create' oprations
+    this.activePrimary = BoshDirectorClient.getActivePrimary();
+    // oobDirectorConfigs - directors for OutOfBand backup scenario
+    this.oobDirectorConfigs = BoshDirectorClient.getOobDirectorConfigs();
+    this.populateCache();
+  }
+
+  clearCache() {
+    this.cache = {};
+  }
+
+  static getInfrastructure() {
+    return BoshDirectorClient.getActivePrimary()[0].infrastructure;
+  }
+
+  static getPrimaryConfigs() {
+    return _.filter(config.directors, function (director) {
+      return director.primary;
+    });
+  }
+
+  static getActivePrimary() {
+    return _.filter(config.directors, function (director) {
+      return director.primary && director.support_create;
+    });
+  }
+
+  static getOobDirectorConfigs() {
+    return _.filter(config.directors, function (director) {
+      return !director.primary;
+    });
+  }
+
+  getConfigByName(name) {
+    return _.head(_.filter(config.directors, (director) => director.name === name));
+  }
+
+  populateCache() {
+    this.clearCache();
+    return Promise
+      .map(config.directors, directorConfig => {
+        return this
+          .getDeploymentsByConfig(directorConfig)
+          .tap(deployments => _.map(deployments, deployment => this.cache[deployment.name] = directorConfig));
```

Comment:
Put a check here for `this.cache[deployment.name] === undefined` and only if undefined then you update the cache.. if this is already cached, then against the entry store an error code something like `this.cache[deployment.name] = CONST.ERR_CODES.DEPLOYMENT_NAME_DUPLICATED;`
This value could be checked in method `getDirectorConfig` and throw the same error.